### ISSUE TITLE
Can't load hdf5 files when using tf.keras.SequentialModel.add() and ModelCheckpoint()

### ISSUE THREAD

**Issue by ddurham2** [tensorflow/tensorflow#20073]
Bug.  Compiled Tensorflow 1.8 and 1.9rc0 both fail. Linux Mint 18.2.  Python-3.5.2  CUDA 9.0/CuDNN 7.0

Have I written custom code: I've attached a file
OS Platform and Distribution: Linux Mint 18.2
TensorFlow installed from: source
TensorFlow version: 1.8 and 1.9rc0
Bazel version: 0.13.1
CUDA/cuDNN version: 9.0/7.0
GPU model and memory: Nvidia 1080Ti
Exact command to reproduce: run the attached script.

When using tf.keras.SequentialModel to build a network, and using tf.keras.callbacks.ModelCheckpoint to save the model/weights during training, the saved hdf5 file cannot be loaded into the same model.

I believe the same problem was observed in keras-1.0
See here: https://github.com/keras-team/keras/issues/2281
And reportedly fixed here: https://github.com/keras-team/keras/commit/1206120d1084cbe45dc2876f002cb572a97e3844

I've attached a minimal script which will readily reproduce the problem here: [save-bug.txt](https://github.com/tensorflow/tensorflow/files/2107894/save-bug.txt)

The output error I receive when attempting to load the model weights into the same model which I saved them from is:

```
Traceback (most recent call last):
  File "./save-bug.py", line 38, in <module>
    predict()
  File "./save-bug.py", line 31, in predict
    model.load_weights('test.hdf5');
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/keras/_impl/keras/engine/network.py", line 1190, in load_weights
    saving.load_weights_from_hdf5_group(f, self.layers)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/keras/_impl/keras/engine/saving.py", line 697, in load_weights_from_hdf5_group
    ' layers.')
ValueError: You are trying to load a weight file containing 1 layers into a model with 0 layers.
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by ddurham2** [comment]
updated description

**Comment by angersson** [comment]
@fchollet Can you take a look at this?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @fchollet, @angersson: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by maelp** [comment]
I have a similar error with TF 1.9 and some models (containing Lambda and Conv2D layers), but other models seem to load correctly

**Comment by maelp** [comment]
The weights are in the HDF5 file, and are saved correctly at https://github.com/tensorflow/tensorflow/blob/r1.9/tensorflow/python/keras/engine/saving.py#L145 (I added logs in the load / save) but when loading from https://github.com/tensorflow/tensorflow/blob/r1.9/tensorflow/python/keras/engine/saving.py#L716 they appear empty

**Comment by tensorflowbutler** [comment]
Nagging Assignees @fchollet, @angersson: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by joba01** [comment]
Same for me, tf.keras models saved and loaded fine, when I added
a reshape layer they still saved but I can't load them anymore.

Model:
`model.add(tf.keras.layers.Reshape([76832]))`

Of course this is just flattening the input and someone can argue that tf.keras.layers.Flatten() does the same thing, but this is just needed because of a TensorRT Bug so reshape is my only option here.

Hope that somebody can take a closer look into that.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by emikulic** [comment]
This is still happening with tf 1.11.0 keras 2.1.6-tf.

Repro:

```python
FN = 'tmp.model.h5'
model = keras.Sequential([
  keras.layers.InputLayer(input_shape=(8, 8, 1)),
  keras.layers.Conv2D(1, 5),
])
keras.models.save_model(model, FN)
keras.models.load_model(FN)
```

This errors out with:

```
Traceback (most recent call last):
  File "./saveload.py", line 24, in <module>
    keras.models.load_model(FN)
  File "tensorflow/python/keras/engine/saving.py", line 233, in load_model
    load_weights_from_hdf5_group(f['model_weights'], model.layers)
  File "tensorflow/python/keras/engine/saving.py", line 784, in load_weights_from_hdf5_group
    ' layers.')
ValueError: You are trying to load a weight file containing 1 layers into a model with 0 layers.
```

It appears to work without the InputLayer, e.g.:

```python
model = keras.Sequential([                                                        
  keras.layers.Conv2D(1, 5, input_shape=(8, 8, 1)),                               
])                                                                                
```

**Comment by emikulic** [comment]
In the working case, `model.layers[0].get_config()` has `'batch_input_shape': (None, 8, 8, 1)`

In the non working case, `batch_input_shape` is absent.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by C-Pro** [comment]
For those still struggling, here is my workaround - instead of saving whole model, i just save and load weights. Obviously in this case you need exact same model defined in learning and inference code.
E.g.:
```python
def makeModel(env, shape, fname, lr):

    # does not work due to https://github.com/tensorflow/tensorflow/issues/20073
    # if os.path.isfile(fname):
    #    print("Loading model")
    #    return load_model(fname)

    m = Sequential()
    m.add(InputLayer(input_shape=shape+(4,)))
    m.add(Conv2D(32, kernel_size=8, strides=4, activation='relu'))
    m.add(Conv2D(64, kernel_size=4, strides=2, activation='relu'))
    m.add(Conv2D(64, kernel_size=3, activation='relu'))
    m.add(Dense(512, activation='relu', activity_regularizer="l2"))
    m.add(Flatten())
    m.add(Dense(env.action_space.n, activation="linear"))
    m.compile(loss=huber_loss, optimizer=tf.keras.optimizers.RMSprop(lr))
    if os.path.isfile(fname):
       m.load_weights(fname)
    return m
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
This is fixed with latest tf-nightly build version '1.15.0-dev20190808'. Thanks!
